srml/offences/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+		/// a vector of offending authorities.
+		///
+		/// It's important that we store all authorities reported for an offence for any kind and
+		/// timeslot since the slashing will increase based on the length of this vec.
+		OffenceReports get(offence_reports):
+			double_map Kind, blake2_256((SessionIndex, TimeSlot))
+			=> Vec<OffenceDetails<T::AccountId, T::IdentificationTuple>>;
+	}
+}
+
+decl_event!(
+	pub enum Event {
+		/// There is an offence reported of the given `kind` happened at the `session_index` and
+		/// (kind-specific) time slot. This event is not deposited for duplicate slashes.
+		Offence(Kind, SessionIndex, TimeSlot),
+	}
+);
+
+decl_module! {
+	/// Offences module, currently just responsible for taking offence reports.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+	}
+}
+
+impl<T: Trait, O: Offence<T::IdentificationTuple>> ReportOffence<T::AccountId, T::IdentificationTuple, O>
+	for Module<T> where
+	T::IdentificationTuple: Clone,
+{
+	fn report_offence(reporter: Option<T::AccountId>, offence: O) {
+		let offenders = offence.offenders();
+		let time_slot = offence.time_slot();
+		let session = offence.session_index();
+		let validator_set_count = offence.validator_set_count();
+
+		// Check if an offence is already reported for the offender authorities
+		// and otherwise stores that report.
+		let mut changed = false;
+		let all_offenders = <OffenceReports<T>>
+			::mutate(&O::ID, &(session, time_slot), |offending_authorities| {
+				for offender in offenders {
+					// TODO [slashing] This prevents slashing for multiple reports of the same kind at the same slot,
+					// note however that we might do that in the future if the reports are not exactly the same (dups).
+					// De-duplication of reports is tricky though, we need a canonical form of the report
+					// (for instance babe equivocation can have headers swapped).
+					if !offending_authorities.iter().any(|details| details.offender == offender) {
+						changed = true;
+						offending_authorities.push(OffenceDetails {
+							offender,
+							count: 0,
+							reporters: reporter.clone().into_iter().collect(),
+						});
+					}
+				}
+
+				if changed {
+					for details in offending_authorities.iter_mut() {
+						details.count += 1;
+					}
+				}
+
+				offending_authorities.clone()
+			});
+
+		if !changed {
+			// The report contained only duplicates, so there is no need to slash again.
+			return
+		}
+
+		// The report is not a duplicate. Deposit an event.
+		Self::deposit_event(Event::Offence(O::ID, session, time_slot));
+
+		let offenders_count = all_offenders.len() as u32;
+		let expected_fraction = O::slash_fraction(offenders_count, validator_set_count);
+		let slash_perbil = all_offenders
+			.iter()
+			.map(|details| {
+				if details.count > 1 {
+					let previous_fraction = O::slash_fraction(
+						offenders_count.saturating_sub(details.count - 1),
```

Comment:
Is this actually the previous fraction of slashing?

`details.count` is initialized to `0` and then incremented to `1` the first time the authority is slashed. But the first slashing is based on `offenders_count`, not on `details.count`. This line seems to assume that the prior slash can be computed based on the `details.count`, but really what we want to store is the number of offenders the last time this offender was slashed.